runtime/compiler/il/J9MethodSymbol.cpp

```diff
@@ -450,6 +450,97 @@ J9::MethodSymbol::safeToSkipArrayStoreChecks()
    return false;
    }
 
+static TR::RecognizedMethod canSkipNonNullableArrayNullStoreCheck[] = {
+   TR::java_lang_invoke_CollectHandle_invokeExact,
+   TR::java_util_ArrayList_add,
+   TR::java_util_ArrayList_ensureCapacity,
+   TR::java_util_ArrayList_get,
+   TR::java_util_ArrayList_remove,
+   TR::java_util_ArrayList_set,
+   TR::java_util_Hashtable_get,
+   TR::java_util_Hashtable_put,
+   TR::java_util_Hashtable_clone,
+   TR::java_util_Hashtable_putAll,
+   TR::java_util_Hashtable_rehash,
+   TR::java_util_Hashtable_remove,
+   TR::java_util_Hashtable_contains,
+   TR::java_util_Hashtable_getEntry,
+   TR::java_util_Hashtable_getEnumeration,
+   TR::java_util_Hashtable_elements,
+   TR::java_util_HashtableHashEnumerator_hasMoreElements,
+   TR::java_util_HashtableHashEnumerator_nextElement,
+   TR::java_util_HashMap_rehash,
+   TR::java_util_HashMap_analyzeMap,
+   TR::java_util_HashMap_calculateCapacity,
+   TR::java_util_HashMap_findNullKeyEntry,
+   TR::java_util_HashMap_get,
+   TR::java_util_HashMap_getNode,
+   TR::java_util_HashMap_putImpl,
+   TR::java_util_HashMap_findNonNullKeyEntry,
+   TR::java_util_HashMap_resize,
+   TR::java_util_HashMap_prepareArray,
+   TR::java_util_HashMap_keysToArray,
+   TR::java_util_HashMap_valuesToArray,
+   TR::java_util_HashMapHashIterator_nextNode,
+   TR::java_util_HashMapHashIterator_init,
+   TR::java_util_Vector_addElement,
+   TR::java_util_Vector_contains,
+   TR::java_util_Vector_subList,
+   TR::java_util_concurrent_ConcurrentHashMap_addCount,
+   TR::java_util_concurrent_ConcurrentHashMap_tryPresize,
+   TR::java_util_concurrent_ConcurrentHashMap_transfer,
+   TR::java_util_concurrent_ConcurrentHashMap_fullAddCount,
+   TR::java_util_concurrent_ConcurrentHashMap_helpTransfer,
+   TR::java_util_concurrent_ConcurrentHashMap_initTable,
+   TR::java_util_concurrent_ConcurrentHashMap_tabAt,
+   TR::java_util_concurrent_ConcurrentHashMap_casTabAt,
+   TR::java_util_concurrent_ConcurrentHashMap_setTabAt,
+   TR::java_util_concurrent_ConcurrentHashMap_TreeBin_lockRoot,
+   TR::java_util_concurrent_ConcurrentHashMap_TreeBin_contendedLock,
+   TR::java_util_concurrent_ConcurrentHashMap_TreeBin_find,
+   TR::java_util_TreeMap_all,
+   TR::java_util_EnumMap_put,
+   TR::java_util_EnumMap_typeCheck,
+   TR::java_util_EnumMap__init_,
+   TR::java_util_EnumMap__nec_,
+   TR::java_util_HashMap_all,
+   TR::java_util_ArrayList_all,
```

Comment:
> I think java/util/ArrayList.toArray([Ljava/lang/Object;)[Ljava/lang/Object; might be a problem here. If the component type of the array argument is a null-restricted value class and is larger than necessary, the method will attempt to store null references in the extra array elements.

During my implementation, I also noticed the implementation of `ArrayList.toArray` on storing `null` into `a[size]`. My reasoning is that `ArrayList.toArray` will have to be updated eventually to work with null-restricted array comment type in this case since it explicitly stores null into an array. Otherwise an exception is thrown and the JCL simply is not workable with the applications using null restricted arrays. 

For now, I can comment out `TR::java_util_ArrayList_all` from `canSkipNonNullableArrayNullStoreCheck[]` with a note so that we can come back to enable it once `ArrayList.toArray` is updated.

Reply:
Addressed in 066df6b